packages/backend-common/src/service/lib/ServiceBuilderImpl.ts

```diff
@@ -120,20 +142,95 @@ export class ServiceBuilderImpl implements ServiceBuilder {
         reject(e);
       });
 
-      const server = stoppable(
-        app.listen(port, host, () => {
-          logger.info(`Listening on ${host}:${port}`);
-        }),
-        0,
-      );
+      let server: http.Server;
+
+      if (httpsSettings) {
+        logger.info('Initializing https server');
+
+        const credentials: { key: string; cert: string } = {
+          key: '',
+          cert: '',
+        };
+        const signingOptions: any = httpsSettings?.certificate;
+
+        if (signingOptions?.algorithm !== undefined) {
+          logger.info('Generating self-signed certificate with attributes');
+
+          const certificateAttributes: Array<any> = [];
+
+          Object.entries(signingOptions?.attributes).map(([name, value]) => {
+            certificateAttributes.push({
+              name: name,
+              value: value,
+            });
+          });
+
+          // TODO: Create a type def for selfsigned.
+          const signatures = require('selfsigned').generate(
+            certificateAttributes,
+            {
+              keySize: signingOptions?.size || 2048,
+              algorithm: signingOptions?.algorithm || 'sha256',
+              days: signingOptions?.days || 30,
+            },
+          );
+
+          logger.info(
+            'Bootstrapping key and cert from self-signed certificate',
+          );
+
+          credentials.key = signatures.private;
+          credentials.cert = signatures.cert;
+        } else {
+          if (fs.existsSync(signingOptions?.key)) {
+            if (fs.lstatSync(signingOptions?.key).isFile()) {
+              logger.info('Bootstrapping key from file');
+
+              credentials.key = fs.readFileSync(signingOptions?.key).toString();
```

Comment:
This will load the key file relative to the working directory, which can be quite unpredictable. For example, an `app-config.yaml` in the root of the repo that specifies `key: https-key.pem`, would actually refer to the path `'packages/backend/https-key.pem'` if you're running the backend as usual.

I think it's better if the path is readable and makes sense as viewed in the `app-config.yaml`, and that paths are relative to the config files themselves. The clean way to do that atm is to just let the full key be configured, and pull in the files using the build-in secret support in the config, e.g. `$secret: file: https-key.pem`

Reply:
To be honest if people enter a relative file URL they should be aware that it will be relative to the host, unless they specify a absolute URL in which case it should resolve if the host has access to it. Regardless of what we make the "relative root" it will always confuse people, some will expect it to be the "logical root" (plugin root folder), others will expect it to be the mono repo root... However in my view forcing a standalone node module to rely on anything other then its "logical root" could create coupling between the "core system" and the "plugin system" which could potentially make it harder distributed the backend further down the line?

Reply:
Resolving it relative to the config file completely removes coupling to the packages, and makes it very portable. Imo it's less surprising and you can know what your config will be by just looking at the repo source.

Reply:
It depends on how you look at it I guess. I think the plugins, more specifically standalone processes, should be fully self-contained as this will ease the process of "chopping up the backend". This would probably require our cfg system to support nested cfgs, chopping up our root config and tweeking some cfg loaders. But I think the end result would make it easier for non-devs to identify the different logical processes that make up the application and deploy them independently without bringing everything and the kitchen sink :p

Reply:
Config would still be a part of that chopped up backend, so it's not really related to what we're discussing